### ISSUE TITLE
Running the root saga only once

### ISSUE THREAD

**Issue by bjoluc** [bmealhouse/next-redux-saga#62]
Hi everyone,
I took a look at #59 and #61. From what I understand, next-redux-wrapper creates three stores per request:

1. On the server: During the app's getInitialProps (the state can be modified here and will be serialized in the end)
2. On the server, during SSR: The state from 1. is used to provide a store for SSR – state changes made during SSR will not be sent to the client, only the serialized state from 1.
3. On the client side using the serialized state from 1.

Let's call the lifetimes of these stores phases and enumerate them as well.

The current solution in the readme modifies the state in phase 1. (`if(!store.getState().saga.ran)`) to remember the saga was run on the server once. The modified state is used in phase 2 (on a different store), which is why `store.getState().saga.ran` is `true` then. This means the rootSaga will not be run in phase 2, which should be perfectly fine as we cannot persistently modify state in phase 2 anyway.  However, the state from phase 1 is also used in phase 3 on the client side, which would lead to the rootSaga not being run on the client-side... :confused:

@bbortt I am wondering if the client-side store actually runs the rootSaga in your setup. Maybe I am getting something wrong here, but as far as I know, the state in phase 3 is the deserialized state from phase 1...

Luckily, within `configureStore` we can take advantage of the second parameter to decide which phase we are in:

```function configureStore(preloadedState, {isServer, req = null}) {```

1. `isServer` is `true` and a request object is provided (`req` is not `null`)
2. `isServer` is `true` and `req` is `null`
3. `isServer` is `false`

Hence, we should be able to achieve the same thing, without adding custom state to the store, like this:

```
  if (req || !isServer) {
    store.sagaTask = sagaMiddleware.run(rootSaga)
  }
```

This works out nicely for me and also runs the saga on the client side.
@bmealhouse @bbortt Is there something I am missing? If not, I would be glad to create a PR for this! :relaxed:

**Comment by bbortt** [comment]
Aight, that might be. I've shortly tested the commit I once tagged and the saga seems indeed not to start on the client side. Seems like I haven't tested it enough as I've migrated away from a custom server to static pages. Which is actually, what I would recommend to everyone.
But I'd be happy to accept your pr ;)

**Comment by bbortt** [comment]
Released in `v4.1.1`. Thanks one last time @bjoluc ;)

**Comment by cddumanov** [comment]
I have weird state overrides. Is it possible that it caused because I have multiple root sagas?

For example here is the start of fetch doctors action. As you see state was changed for doctor isLoading: false to isLoading: true
![image](https://user-images.githubusercontent.com/43003794/74384023-130c7380-4df9-11ea-883b-f1994d1a1c42.png)
After that fetch cities action was started right before. And isLoading for doctors was changed back to false.
![image](https://user-images.githubusercontent.com/43003794/74384053-23bce980-4df9-11ea-9c75-37b3fb481be2.png)
![image](https://user-images.githubusercontent.com/43003794/74384070-2c152480-4df9-11ea-8602-47330b074ed1.png)
![image](https://user-images.githubusercontent.com/43003794/74384087-30414200-4df9-11ea-9a84-c2644506e01c.png)

For more info please see https://stackoverflow.com/questions/60158683/react-redux-nextjs-weird-state-race-overrides

**Comment by bbortt** [comment]
*Quick side note: Please open a new Issue next time. This one was closed and your problem is not really related to it.*

Anyways: I guess the problem probably lies in your reducers. You need to return the currenct `state` rather than the `initialState`. If the `state` was `null` or `undefined`, `initalState` would be assigned by the definition of your reducer. Because state is immutable you need to splice and return `state` at all time. 
(Actually you would not need to splice `state` if there was no change, but in my opinion it's good manner to do so anyways.)
*Never* `return initialState` - this loses your current `state`.
Construct your reducer like this:
```js
export const cityReducer = (state = initialState /* this makes sure `state` is never `null` or `undefined` */, action) => {
  switch (action.type) {
    case actionTypes.FETCH_CITIES:
// -      return initialState
      return {...state}
    case actionTypes.FETCH_CITIES_SUCCESS:
      return {
        ...state,
        cities: action.data
      }
    case actionTypes.FETCH_CITIES_FAILURE:
// -      return initialState
      return {...state}
    default:
// -      return initialState
      return {...state}
  }
}
```
Let me know if this helped.

**Comment by cddumanov** [comment]
@bbortt You pointed to the exact problem. Thanks a lot!
